nupic/research/frameworks/pytorch/imagenet/imagenet_experiment.py

```diff
@@ -869,47 +894,173 @@ def get_free_port(self):
+            expand_result_to_time_series=[exp + ": validation results"],
+            stop_experiment=[exp + ".stop_experiment"]
         )
 
 
+class ContinualLearningExperiment(ContinualLearningMetrics, SupervisedExperiment):
+
+    def setup_experiment(self, config):
+
+        super().setup_experiment(config)
+        # override epochs to validate to not validate within the inner loop over epochs
+        self.epochs_to_validate = []
+
+        self.current_task = 0
+
+        # defines how many classes should exist per task
+        self.num_tasks = config.get("num_tasks", 1)
+
+        self.num_classes = config.get("num_classes", None)
+        assert self.num_classes is not None, "num_classes should be defined"
+
+        self.num_classes_per_task = math.floor(self.num_classes / self.num_tasks)
+
+        # applying target transform depending on type of CL task
+        # task - we know the task, so the network is multihead
+        # class - we don't know the task
+        self.cl_experiment_type = config.get("cl_experiment_type", "class")
+        if self.cl_experiment_type == "task":
+            # override the target transform
+            self.dataset_args["target_transform"] = (
+                transforms.Lambda(lambda y: y % self.num_classes_per_task)
+            )
+
+        # whitelist evaluation metrics
+        self.evaluation_metrics = config.get(
+            "evaluation_metrics", ["eval_all_visited_tasks"]
+        )
+        for metric in self.evaluation_metrics:
+            if not hasattr(self, metric):
+                raise ValueError(f"Metric {metric} not available.")
+
+    def should_stop(self):
+        """
+        Whether or not the experiment should stop. Usually determined by the
+        number of epochs but customizable to any other stopping criteria
+        """
+        return self.current_task >= self.num_tasks
+
+    def run_task(self):
+        """Run outer loop over tasks"""
+        # configure the sampler to load only samples from current task
+        print("Training...")
+        self.train_loader.sampler.set_active_tasks(self.current_task)
+
+        # run epochs, inner loop
+        # TODO: return the results from run_epoch
+        for _ in range(self.epochs):
+            self.run_epoch()
+
+        ret = {}
+        for metric in self.evaluation_metrics:
+            eval_function = getattr(self, metric)
+            temp_ret = eval_function()
+            print(temp_ret)
+            for k, v in temp_ret.items():
+                ret[f"{metric}__{k}"] = v
+
+        # TODO: fix this
+        ret.update(
+            timestep_begin=0,
+            timestep_end=1,
+            learning_rate=self.get_lr()[0],
+            extra_val_results=[],
+        )
+
+        self.current_task += 1
+        return ret
+
+    @classmethod
+    def aggregate_results(cls, results):
+        """Run validation in single GPU"""
+        print("Aggregated results")
```

Comment:
You should check for all your print statements

Reply:
Should I just change all print statements to logger.debug instead, what do you think? This is still in development, I don't want to be removing print statements and putting them back everytime we push a new PR on this (actually I've already removed 90% of the print statements we had).

Reply:
Part of the process could be simply not committing your print statement as well. I think we should try and keep master as tidy as possible.